File: src/opengl/texture.rs
use gl33::{global_loader::{glBindTexture, glGenTextures, glGenerateMipmap, glTexImage2D}, GL_RGB, GL_TEXTURE_2D, GL_UNSIGNED_BYTE};
use image::{RgbImage};

use crate::opengl::types::GLuint;

pub struct Texture {
  pub name: GLuint,
}

impl Texture {
  pub fn new() -> Option<Self> {
    let mut name = 0;
    unsafe { glGenTextures(1, &mut name) }

    if name == 0 {
      return None;
    }

    return Some(Self { name });
  }

  pub fn bind(&self) {
    unsafe { glBindTexture(GL_TEXTURE_2D, self.name) }
  }

  pub fn load_RGB(&self, image: RgbImage) {
    unsafe {
      glTexImage2D(
        GL_TEXTURE_2D,
        0,
        GL_RGB.0 as i32,
        image.width() as i32,
        image.height() as i32,
        0,
        GL_RGB,
        GL_UNSIGNED_BYTE,
        image.as_ptr() as *const _,
      );
      glGenerateMipmap(GL_TEXTURE_2D);
    };
  }
}

File: opengl/texture.py
import cv2
import OpenGL.GL as GL


class Texture:
    def __init__(self):
        self.texture = GL.glGenTextures(1)

    def bind(self, index: int):
        GL.glActiveTexture(GL.GL_TEXTURE0 + index)
        GL.glBindTexture(GL.GL_TEXTURE_2D, self.texture)

    def load(self, filename):
        image = cv2.imread(filename)
        GL.glTexImage2D(
            GL.GL_TEXTURE_2D,
            0,
            GL.GL_RGB,
            image.shape[1],
            image.shape[0],
            0,
            GL.GL_BGR,
            GL.GL_UNSIGNED_BYTE,
            image
        )
        GL.glGenerateMipmap(GL.GL_TEXTURE_2D)
